diff --git a/google/storage/v2/storage.proto b/google/storage/v2/storage.proto
index c9c027345..0bcd970e0 100644
--- a/google/storage/v2/storage.proto
+++ b/google/storage/v2/storage.proto
@@ -1177,6 +1171,22 @@ message BidiReadObjectRedirectedError {
   optional string routing_token = 2;
 }
 
+// Error proto containing details for a redirected write. This error is only
+// returned on initial open in case of a redirect.
+message BidiWriteObjectRedirectedError {
+  // The routing token that should be used when reopening the write stream.
+  optional string routing_token = 1;
+
+  // Opaque value describing a previous write.
+  optional BidiWriteHandle write_handle = 2;
+
+  // The generation of the object that triggered the redirect.
+  // Note that if this error was returned as part of an appendable object
+  // create, this object generation is now successfully created and
+  // append_object_spec should be used when reconnecting.
+  optional int64 generation = 3;
+}
+
 // Error extension proto containing details for all outstanding reads on the
 // failed stream
 message BidiReadObjectError {
@@ -1246,6 +1262,14 @@ message BidiReadHandle {
   bytes handle = 1 [(google.api.field_behavior) = REQUIRED];
 }
 
+// BidiWriteHandle contains a handle from a previous BidiWriteObject
+// invocation. The client can use this as an optimized way of opening subsequent
+// bidirectional streams to the same object.
+message BidiWriteHandle {
+  // Opaque value describing a previous write.
+  bytes handle = 1;
+}
+
 // Describes an attempt to insert an object, possibly over multiple requests.
 message WriteObjectSpec {
   // Required. Destination object, including its name and its metadata.
@@ -1283,6 +1307,12 @@ message WriteObjectSpec {
   // you must start the upload over from scratch, this time sending the correct
   // number of bytes.
   optional int64 object_size = 8;
+
+  // If true, the object will be created in appendable mode.
+  // This field may only be set when using BidiWriteObject.
+  optional bool appendable = 9;
+
+  reserved 2;
 }
 
 // Request message for WriteObject.
@@ -1352,6 +1382,38 @@ message WriteObjectResponse {
   }
 }
 
+// Describes an attempt to append to an object, possibly over multiple requests.
+message AppendObjectSpec {
+  // The name of the bucket containing the object to write.
+  string bucket = 1 [
+    (google.api.resource_reference).type = "storage.googleapis.com/Bucket"
+  ];
+
+  // The name of the object to open for writing.
+  string object = 2 [
+    (google.api.field_behavior) = REQUIRED
+  ];
+
+  // The generation number of the object to open for writing.
+  int64 generation = 3 [(google.api.field_behavior) = REQUIRED];
+
+  // Makes the operation conditional on whether the object's current
+  // metageneration matches the given value.
+  optional int64 if_metageneration_match = 4;
+
+  // Makes the operation conditional on whether the object's current
+  // metageneration does not match the given value.
+  optional int64 if_metageneration_not_match = 5;
+
+  // An optional routing token that influences request routing for the stream.
+  // Must be provided if a BidiWriteObjectRedirectedError is returned.
+  optional string routing_token = 6;
+
+  // An optional write handle returned from a previous BidiWriteObjectResponse
+  // message or a BidiWriteObjectRedirectedError error.
+  optional BidiWriteHandle write_handle = 7;
+}
+
 // Request message for BidiWriteObject.
 message BidiWriteObjectRequest {
   // The first message of each stream should set one of the following.
@@ -1363,6 +1434,9 @@ message BidiWriteObjectRequest {
     // For non-resumable uploads. Describes the overall upload, including the
     // destination bucket and object name, preconditions, etc.
     WriteObjectSpec write_object_spec = 2;
+
+    // For appendable uploads. Describes the object to append to.
+    AppendObjectSpec append_object_spec = 11;
   }
 
   // Required. The offset from the beginning of the object at which the data
@@ -1435,6 +1509,11 @@ message BidiWriteObjectResponse {
     // the upload has finalized.
     Object resource = 2;
   }
+
+  // An optional write handle that will periodically be present in response
+  // messages. Clients should save it for later use in establishing a new stream
+  // if a connection is interrupted.
+  optional BidiWriteHandle write_handle = 3;
 }
 
 // Request message for ListObjects.
